[lc709203f] Add documentation for new component (#4800)
OttoWinter/esphomedocs: components/sensor/lc709203f.rst

(`datasheet <https://cdn-learn.adafruit.com/assets/assets/000/094/597/original/LC709203F-D.PDF>`__) 
battery monitor with ESPHome. This device is available as a 
`standalone sensor <https://www.adafruit.com/product/4712>`__ and it is also one of the battery
monitor chips used on the `ESP32 Feather <https://www.adafruit.com/product/5477>`__ dev boards.

.. image:: images/LC709203f_product_image.jpg
    :width: 30.0%

.. image:: images/LC709203f_feather_product_image.jpg
    :width: 30.0%

.. note:: 

    This device does not contain a temperature sensor. Only enable the temperature sensor option 
    if you have attached a thermistor as shown in :ref:`temperature-sensor`.

.. code-block:: yaml

    # Example configuration entry
    sensor:
      - platform: lc709203f
        size: 2000
        voltage: 3.7
        battery_voltage:
          name: "Battery Voltage"
        battery_level:
          name: "Battery"
        temperature:
          name: "Pack Temperature"
          thermistor_b_constant: 0xA5A5

Configuration variables:
-----------------------------------------

- **size** (*Optional*): Size of the battery in mAH. 

>   - Valid values are integers between 200 and 3000.

Reviewer: @ilikecake it looks like this should have 100 as the minimum. Can you confirm?
```suggestion
  - Valid values are integers between 100 and 3000.
```

Author: Yeah, this is right.

>   - Defaults to 500 mAH. You should really set this correctly but if you really don't know, this default is probably okay.

Reviewer: ```suggestion
  - Defaults to 500 mAH; we highly recommend that this value is set as appropriate for your battery pack.
```

>   - see :ref:`pack-size-and-voltage` for help if you don't know your pack size.

Reviewer: ```suggestion
  - See :ref:`pack-size-and-voltage` for help if you don't know your pack size.
```

  
- **voltage** (*Optional*): nominal voltage of the battery pack in V.

  - Valid values are ``3.7`` or ``3.8``
  - Defaults to ``3.7``. This is the correct value for the Adafruit batteries.
  - see :ref:`pack-size-and-voltage` for help if you don't know your pack voltage.

>   - see :ref:`pack-voltage` for more information on how this value is used.

Reviewer: ```suggestion
  - See :ref:`pack-size-and-voltage` for help if you don't know your pack voltage.
  - See :ref:`pack-voltage` for more information on how this value is used.
```

LC709203F Battery Monitor
===============================================

.. seo::
    :description: Instructions for setting up LC709203F battery monitor.
    :image: lc709203f.jpg
    :keywords: LC709203F

The ``lc709203f`` sensor platform allows you to use a LC709203F
(`datasheet <https://cdn-learn.adafruit.com/assets/assets/000/094/597/original/LC709203F-D.PDF>`__) 
battery monitor with ESPHome. This device is available as a 
`standalone sensor <https://www.adafruit.com/product/4712>`__ and it is also one of the battery

> monitor chips used on their `ESP32 Feather <https://www.adafruit.com/product/5477>`__ dev boards.

Reviewer: ```suggestion
monitor chips used on the `ESP32 Feather <https://www.adafruit.com/product/5477>`__ dev boards.

The :ref:`I²C <i2c>` component is required to be set up in your configuration.
```

.. image:: images/LC709203f_product_image.jpg
    :width: 30.0%

.. image:: images/LC709203f_feather_product_image.jpg
    :width: 30.0%

.. note:: 

    This device does not contain a temperature sensor. Only enable the temperature sensor option 
    if you have attached a thermistor as shown in :ref:`temperature-sensor`.

> Prerequisites:
> -----------------------------------------
> 
> This component relies on the :ref:`I²C <i2c>` componenent to be setup. See the reference for
> that component for details. A basic example of what that looks like is below.
> 
> .. code-block:: yaml
> 
>     i2c:
>       sda: SDA
>       scl: SCL
>       scan: true
>       id: bus_a
> 
> Configuration Example:
> -----------------------------------------
> An example configuration YAML is shown below.
> 

Reviewer: Please remove this.
```suggestion
```

.. code-block:: yaml

>     sensor:

Reviewer: ```suggestion
    # Example configuration entry
    sensor:
```

      - platform: lc709203f
        size: 2000
        voltage: 3.7
        battery_voltage:
          name: "Battery Voltage"
        battery_level:
          name: "Battery"
        temperature:
          name: "Pack Temperature"
          thermistor_b_constant: 0xA5A5

Configuration variables:
-----------------------------------------

- **size** (*Optional*): Size of the battery in mAH. 

  - Valid values are integers between 200 and 3000.

>   - Defaults to 500 mAH. You should really set this correctly but if you really don't know, this default is probably okay.

Reviewer: I don't really think we should have a default for this. If you blow up your battery, it's entirely on you.

Author: If you want, I can change this. However, since this sensor only uses voltage to guess at the charge level of the battery, in theory, setting this incorrectly should still give a roughly correct reading. I figured that having a default for this would let the user get *something* from the sensor without knowing battery size. I could go either way on this one.

Reviewer: Cool -- I don't have a terribly strong opinion on it, myself. 😅 Curious to get the input of others who know more about these things than I do.

  - see :ref:`pack-size-and-voltage` for help if you don't know your pack size.
  
- **voltage** (*Optional*): nominal voltage of the battery pack in V.

  - Valid values are ``3.7`` or ``3.8``
  - Defaults to ``3.7``. This is the correct value for the Adafruit batteries.
  - see :ref:`pack-size-and-voltage` for help if you don't know your pack voltage.
  - see :ref:`pack-voltage` for more information on how this value is used.

> - **battery_voltage** (*Optional*): The setup information for the voltage sensor.
> 
>   - Standard options from :ref:`Sensor <config-sensor>`.
>   - **name** (*Optional*): I suggest setting this one.

Reviewer: ```suggestion
- **battery_voltage** (*Optional*): Configuration for the voltage sensor. All options from :ref:`sensor <config-sensor>`.
```

> - **battery_level** (*Optional*): The setup information for the battery level sensor.
> 
>   - Standard options from :ref:`Sensor <config-sensor>`.
>   - **name** (*Optional*): I suggest setting this one.

Reviewer: ```suggestion
- **battery_level** (*Optional*): Configuration for the battery level sensor. All options from :ref:`Sensor <config-sensor>`.
```

> - **temperature** (*Optional*): The setup information for the battery temperature sensor.
>   
>   - **thermistor_b_constant** (*Optional*): The B-constant of the thermistor. Defaults 
>     to `0x0BA6` but if you are using a thermistor, you should really change this.
>   - Standard options from :ref:`Sensor <config-sensor>`.
>   - **name** (*Optional*): I suggest setting this one.

Reviewer: As discussed on the code PR, let's make the B-constant required -- but also, let's not refer to its value in hexadecimal (I alluded to this in my comment on the code PR). I know the chip's datasheet does this, but given the use case, it just plain doesn't make sense to use hexadecimal notation. You're not going to find the B-constant in any thermistor's datasheet written in hexadecimal. 😉

```suggestion
- **temperature** (*Optional*): Configuration for the battery temperature sensor.
  
  - **thermistor_b_constant** (**Required**): The B-constant of the thermistor you are using.
  - All options from :ref:`Sensor <config-sensor>`.
```

- **update_interval** (*Optional*, :ref:`config-time`): The interval to check the
  sensor. Defaults to ``60s``.

Home Assistant View:
-----------------------------------------

When properly set up, the sensor will report values to Home Assistant as shown:

.. image:: images/LC709203f_ui.jpg
    :align: center
    :width: 40.0%

.. _pack-size-and-voltage:

Pack Size and Nominal Voltage
-----------------------------------------

> If you don't know the pack size and nominal voltage of your battery, it is typically written
> on the battery as shown below.

Reviewer: ```suggestion
The pack size and nominal voltage of your battery is typically printed on the battery as shown below.
```

.. image:: images/LC709203f_battery_markings.jpg
    :align: center
    :width: 70.0%

.. _pack-voltage:

Pack Voltage
-----------------------------------------

> When setting up the device. You will need to set the nominal pack voltage of the device. This 
> is used internally to make the sensor more accurate. The nominal voltage is used to set the 
> ``change of the parameter`` register per table 8 in the datasheet.

Reviewer: ```suggestion
You will need to configure your device with the correct nominal pack voltage. This
is used by the IC to improve the sensor's accuracy. The nominal voltage is used to set the 
``change of the parameter`` register per table 8 in the datasheet.
```

.. image:: images/LC709203f_battery_profiles.jpg
    :align: center
    :width: 70.0%

We assume that the device is a ``-01`` or ``-03`` device. This is the correct setup for the
Adafruit sensors and batteries.

.. _temperature-sensor:

Temperature Sensor Information
-----------------------------------------

If you want to measure temperature of the battery, you **must** have a thermistor attached to 
the device as shown in figure 2 of the datasheet.

.. image:: images/LC709203f_thermistor_connection.jpg
    :align: center
    :width: 70.0%

> Special Thanks
> -----------------------------------------

Reviewer: ```suggestion
Acknowledgments
---------------
```
